Add autocompleter from Brave

return42 opened the pull request:
## What does this PR do?

Add autocompleter from Brave / BTW add SPDX tag to `autocomplete.py`

## Why is this change important?

one more alternative

## How to test this PR locally?

    make run

Activate brave auto completion and test completion in various languages

## Related issues

Initial commit was cherry picked from https://github.com/searx/searx/pull/3109 (and slightly modified to fit SearXNG needs)

allendema committed:
[enh] Add autocompleter from Brave

Raw response example: https://search.brave.com/api/suggest?q=how%20to:%20with%20j

Headers are needed in order to get a 200 response, thus Searx user-agent is used.

Other URL param could be  '&rich=false' or  '&rich=true'.

Cherry-pick: https://github.com/allendema/searx/commit/71786bf9cb6fbb175a054692e6951e77769aac1b

return42 committed:
[pylint] searx/autocomplete.py

Fix remarks from pylint, BTW set SPDX-License-Identifier.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/autocomplete.py:
```diff
@@ -36,7 +25,26 @@ def get(*args, **kwargs):
+    results = []
+
+    if resp.ok:
+        data = loads(resp.text)
```
`data = resp.json()` doesn't trigger the charset detection for each request (see the logs).

the `json()` method has shortcut to detect UTF encoding: https://github.com/encode/httpx/blob/321d4aa5097fe7f24cdfed7191c44de589294780/httpx/_models.py#L1514

return42 replied:
> see the logs ..

I debugged ..

```
(Pdb) l
 39  	    results = []
 40  	
 41  	    if resp.ok:
 42  	        import pdb
 43  	        pdb.set_trace()
 44  ->	        data = loads(resp.text)
 45  	        for item in data[1]:
 46  	            results.append(item)
 47  	    return results
 48  	
 49  	
(Pdb) resp.text
DEBUG   charset_normalizer            : override steps (5) and chunk_size (512) as content does not fit (153 byte(s) given) parameters.
INFO    charset_normalizer            : ascii passed initial chaos probing. Mean measured chaos is 0.000000 %
DEBUG   charset_normalizer            : ascii should target any language(s) of ['Latin Based']
INFO    charset_normalizer            : ascii is most likely the one. Stopping the process.
```

what means this logs are not related to json parser.  I assume the messages are related to a typecast of `resp.text` .. but I don't really know what I can do ...

return42 replied:
To avoid such issues from httpx.Response, we need to use httpx.Response.json() / fixed in e9588b70

@dalf: Should we replace `json.load(resp.text)` by `resp.json()` entirely? (I can do it in another PR)

dalf reviewed (changes requested):
Work as expected but 23156f2b692ec110e0140df0767a9f5927018426 is no go for me because it requires manual update of searx/data/brave_languages.json

the right solution is to have `fetch_supported_languages` function for the brave engine.
However, since the brave engine uses the xpath module and inheritance is not implemented, there is no straightforward solution.

So for now, I would
* remove language support
* find a way to specify a `fetch_supported_languages` function the engines using the xpath or json modules.

return42 committed:
[fix] brave autocompleter: charset_normalizer issues

Use httpx.Response.json() to avoid charset_normalizer issues:

DEBUG   charset_normalizer            : override steps (5) and chunk_size (512) as content does not fit (153 byte(s) given) parameters.
INFO    charset_normalizer            : ascii passed initial chaos probing. Mean measured chaos is 0.000000 %
DEBUG   charset_normalizer            : ascii should target any language(s) of ['Latin Based']
INFO    charset_normalizer            : ascii is most likely the one. Stopping the process.

[1] https://www.python-httpx.org/api/#response

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf commented on searx/autocomplete.py:
```diff
@@ -36,7 +26,26 @@ def get(*args, **kwargs):
+    country = 'all'
+    # if lang in _brave:
+    #    country = lang
+    kwargs = {'cookies': {'country': country}}
```
I see this cookie `ui_lang=en-us` but not `country`.

return42 replied:
When I go to https://search.brave.com/ and trace the suggestion request, I see:

![grafik](https://user-images.githubusercontent.com/554536/150690928-9726ff62-d60b-4a5f-90e9-a8a97d85333d.png)

dalf replied:
I see
![image](https://user-images.githubusercontent.com/1594191/150692379-01eeaf35-d81a-43ca-b00d-e3edd1c84d78.png)

may be I use the wrong setting.

dalf reviewed (approved):
I left a comment about the cookie, not important now since the language selection is not supported.

Thank you.
